nvflare/app_common/workflows/wf_controller.py

```diff
@@ -0,0 +1,296 @@
+
+        self.engine = None
+        self.fl_ctx = None
+
+    def start_controller(self, fl_ctx: FLContext):
+        self.fl_ctx = fl_ctx
+        self.log_info(fl_ctx, "Initializing controller workflow.")
+        self.engine = self.fl_ctx.get_engine()
+        self.clients = self.engine.get_clients()
+
+        self.setup_wf_queue()
+
+        self.log_info(fl_ctx, "workflow controller started")
+
+    def setup_wf_queue(self):
+        wf_comm_api = self.find_wf_comm_in_wf()
+        wf_comm_api.set_queue(self.wf_queue)
+        wf_comm_api.set_result_pull_interval(self.comm_msg_pull_interval)
+        wf_comm_api.meta.update({SITE_NAMES: self.get_site_names()})
+
+        print(f"\n \n  {wf_comm_api.meta=}")
+
+    def find_wf_comm_in_wf(self):
+        attr_objs = [getattr(self.wf, attr_name, None) for attr_name in dir(self.wf)]
+        wf_comm_attrs = [attr for attr in attr_objs if isinstance(attr, WFCommAPI)]
+        if wf_comm_attrs:
+            return wf_comm_attrs[0]
+        else:
+            raise RuntimeError(f"missing required attribute with type of 'WFCommAPI' in {self.wf.__class__.__name__}")
+
+    def control_flow(self, abort_signal: Signal, fl_ctx: FLContext):
+        try:
+            future = self._thread_pool_executor.submit(self.ctrl_msg_loop, fl_ctx=fl_ctx, abort_signal=abort_signal)
+            self.wf.run()
+            self.stop_msg_queue("job completed", fl_ctx)
+            future.result()
+        except Exception as e:
+            error_msg = secure_format_traceback()
+            self.log_error(fl_ctx, error_msg)
+            self.system_panic(error_msg, fl_ctx=fl_ctx)
+        finally:
+            wait_time = self.comm_msg_pull_interval + 0.05
+            self.stop_msg_queue("job finished", fl_ctx, wait_time)
+
+    def stop_msg_queue(self, stop_message, fl_ctx, wait_time: float = 0):
+        self.wf_queue.stop(stop_message)
+        self.log_info(fl_ctx, stop_message)
+
+        if wait_time > 0:
+            self.log_info(fl_ctx, f"wait for {wait_time} sec")
+            time.sleep(wait_time)
+
+    def stop_controller(self, fl_ctx: FLContext):
+        self.stop_msg_queue("job completed", fl_ctx)
+        if self._thread_pool_executor:
+            self._thread_pool_executor.shutdown()
+
+    def process_result_of_unknown_task(
+        self, client: Client, task_name: str, client_task_id: str, result: Shareable, fl_ctx: FLContext
+    ):
+        pass
+
+    def ctrl_msg_loop(self, fl_ctx: FLContext, abort_signal: Signal):
+
+        if self.wf_queue is None:
+            raise ValueError("WFQueue must provided")
+
+        try:
+            while True:
+                if abort_signal.triggered:
+                    break
+                if not self.wf_queue.has_ctrl_msg():
+                    time.sleep(self.comm_msg_pull_interval)
+                else:
+                    item = self.wf_queue.get_ctrl_msg()
```

Comment:
From https://docs.python.org/3/library/queue.html#queue.Queue.task_done, we might need to consider when to call this.  Without calling it, it might result in memory leak (even the element is just a one-key dictionary).